src/aiida/manage/configuration/settings.py

```diff
@@ -25,70 +26,86 @@
 DEFAULT_DAEMON_LOG_DIR_NAME = 'log'
 DEFAULT_ACCESS_CONTROL_DIR_NAME = 'access'
 
-# Assign defaults which may be overriden in set_configuration_directory() below
-AIIDA_CONFIG_FOLDER: pathlib.Path = pathlib.Path(DEFAULT_AIIDA_PATH).expanduser() / DEFAULT_CONFIG_DIR_NAME
-DAEMON_DIR: pathlib.Path = AIIDA_CONFIG_FOLDER / DEFAULT_DAEMON_DIR_NAME
-DAEMON_LOG_DIR: pathlib.Path = DAEMON_DIR / DEFAULT_DAEMON_LOG_DIR_NAME
-ACCESS_CONTROL_DIR: pathlib.Path = AIIDA_CONFIG_FOLDER / DEFAULT_ACCESS_CONTROL_DIR_NAME
+__all__ = ('AiiDAConfigPathResolver',)
 
 
-def create_instance_directories() -> None:
+@final
+class AiiDAConfigPathResolver:
+    """Path resolver for getting daemon dir, daemon log dir and access control dir location."""
+
+    _glb_aiida_config_folder: pathlib.Path = pathlib.Path(DEFAULT_AIIDA_PATH).expanduser() / DEFAULT_CONFIG_DIR_NAME
+
+    def __init__(self, config_folder: pathlib.Path | None) -> None:
+        self._aiida_path = config_folder or self._glb_aiida_config_folder
+
+    @property
+    def aiida_path(self) -> pathlib.Path:
+        return self._aiida_path
+
+    @property
+    def daemon_dir(self) -> pathlib.Path:
+        return self._aiida_path / DEFAULT_DAEMON_DIR_NAME
+
+    @property
+    def daemon_log_dir(self) -> pathlib.Path:
+        return self._aiida_path / DEFAULT_DAEMON_DIR_NAME / DEFAULT_DAEMON_LOG_DIR_NAME
+
+    @property
+    def access_control_dir(self) -> pathlib.Path:
+        return self._aiida_path / DEFAULT_ACCESS_CONTROL_DIR_NAME
+
+    @classmethod
+    def get_configuration_directory(cls):
+        """Return the path of the configuration directory."""
+        return cls._glb_aiida_config_folder
```

Comment:
I see what you are doing here, but the fact that this is a classmethod might lead to  surprising behavior no? Consider the following example:
```python
c = AiiDAConfigPathResolver("/some/path")
assert c.get_configuration_directory() == "/some/path"
```
will fail, right?

Essentially you have turned the `AIIDA_CONFIG_FOLDER` global variable into the `AiiDAConfigPathResolver._glb_aiida_config_folder` class variable.

I guess the real problem was not so much the `AIIDA_CONFIG_FOLDER` global variable (because that essentially still exists) but the other derivative global variables. If I understand correctly is that those global vars `DAEMON_DIR`, `DAEMON_LOG_DIR`, etc were computed on import. So if some module imported them, they would be fixed forever and not be updated when `set_configuration_directory` would be called.

Instead of changing all the logic of how `AIIDA_CONFIG_FOLDER` is set, why not just address the real problem of these derivative paths. I think they are only really used in `Profile.filepaths` and as I already mentioned in another comment, I think this logic doesn't really belong here, but I think should be done in some method on `Config` like `Config.get_profile_filepaths()`. This method can then dynamically compute the paths just as you do here in `AiiDAConfigPathResolver`.

Reply:
Good point! I think it makes more sense to keep `AiiDAConfigPathResolver` for getting config path structure and I add `AiiDAConfigDir` with the global attribute and provide setter/getter.